File: untests/sources/untests_high/flipbook_player.cpp
/*******************************************************************************
 * This file is part of the "Enduro2D"
 * For conditions of distribution and use, see copyright notice in LICENSE.md
 * Copyright (C) 2018-2019, by Matvey Cherevko (blackmatov@gmail.com)
 ******************************************************************************/

#include "_high.hpp"
using namespace e2d;

TEST_CASE("flipbook_player") {
    {
        flipbook_player fp;
        REQUIRE(math::approximately(fp.time(), 0.f));
        REQUIRE(math::approximately(fp.speed(), 1.f));
        REQUIRE_FALSE(fp.looped());
        REQUIRE(fp.stopped());
        REQUIRE_FALSE(fp.playing());
        REQUIRE(fp.sequence() == make_hash(""));
    }
    {
        flipbook_player fp;

        fp.playing(true);
        REQUIRE(fp.playing());
        REQUIRE_FALSE(fp.stopped());

        fp.stopped(true);
        REQUIRE_FALSE(fp.playing());
        REQUIRE(fp.stopped());

        fp.stopped(false);
        REQUIRE(fp.playing());
        REQUIRE_FALSE(fp.stopped());

        fp.playing(false);
        REQUIRE_FALSE(fp.playing());
        REQUIRE(fp.stopped());
    }
    {
        flipbook_player fp;

        REQUIRE(&fp.play(5.f) == &fp);
        REQUIRE(math::approximately(fp.time(), 5.f));
        REQUIRE(math::approximately(fp.speed(), 1.f));
        REQUIRE_FALSE(fp.looped());
        REQUIRE_FALSE(fp.stopped());
        REQUIRE(fp.playing());
        REQUIRE(fp.sequence() == make_hash(""));

        REQUIRE(&fp.play("hello") == &fp);
        REQUIRE(math::approximately(fp.time(), 0.f));
        REQUIRE(math::approximately(fp.speed(), 1.f));
        REQUIRE_FALSE(fp.looped());
        REQUIRE_FALSE(fp.stopped());
        REQUIRE(fp.playing());
        REQUIRE(fp.sequence() == make_hash("hello"));

        REQUIRE(&fp.stop(5.f) == &fp);
        REQUIRE(math::approximately(fp.time(), 5.f));
        REQUIRE(math::approximately(fp.speed(), 1.f));
        REQUIRE_FALSE(fp.looped());
        REQUIRE(fp.stopped());
        REQUIRE_FALSE(fp.playing());
        REQUIRE(fp.sequence() == make_hash("hello"));

        REQUIRE(&fp.stop("") == &fp);
        REQUIRE(math::approximately(fp.time(), 0.f));
        REQUIRE(math::approximately(fp.speed(), 1.f));
        REQUIRE_FALSE(fp.looped());
        REQUIRE(fp.stopped());
        REQUIRE_FALSE(fp.playing());
        REQUIRE(fp.sequence() == make_hash(""));
    }
}

File: samples/bin/library/scripts/emmy/components/flipbook_player.lua
---@class flipbook_player
local flipbook_player = {
    ---@type number
    time = 0,

    ---@type number
    speed = 1,

    ---@type boolean
    looped = false,

    ---@type boolean
    stopped = true,

    ---@type boolean
    playing = false,

    ---@type str_hash
    sequence = str_hash.new(),

    ---@type flipbook_asset
    flipbook = nil
}

---@overload fun(self: flipbook_player, time: number)
---@overload fun(self: flipbook_player, sequence: string)
---@overload fun(self: flipbook_player, sequence: str_hash)
---@param self flipbook_player
function flipbook_player.stop(self, ...) end

---@overload fun(self: flipbook_player, time: number)
---@overload fun(self: flipbook_player, sequence: string)
---@overload fun(self: flipbook_player, sequence: str_hash)
---@param self flipbook_player
function flipbook_player.play(self, ...) end

---@type flipbook_player
_G.flipbook_player = _G.flipbook_player or flipbook_player
